# Day 1 – Module 1 & 2: Python for AI Workflows & Understanding AI APIs

**Topics covered:**  
- Python for AI Workflows: structuring scripts, JSON/CSV/API payloads, functions and modular design, `requests`, handling responses and errors, logging and debugging.  
- Understanding AI APIs: REST fundamentals, endpoints, authentication, parameters, API architecture in enterprise AI systems.

**What you will do in this notebook:**  
Build a complete, production-grade Python workflow that calls AI APIs: from loading config and secrets, through building request payloads from real data (CSV, JSON), to sending requests, handling errors with retry logic, and logging every step for traceability. Every pattern here maps directly to how production AI pipelines are built — config separate from code, modular functions, structured logging with correlation IDs, and resilient error handling that keeps batch jobs running when individual calls fail.

## 1. Where your Python code fits — enterprise AI API architecture

```
  ┌──────────────────────────────────────────────────────────────────────────┐
  │                       ENTERPRISE AI API FLOW                            │
  │                                                                          │
  │  ┌──────────────┐    HTTP POST     ┌────────────────┐    Forward     ┌──────────────┐
  │  │  BLOCK A      │  ───────────►   │  BLOCK B        │  ──────────►  │  BLOCK C      │
  │  │  Your Python  │  JSON + Auth    │  API Gateway /  │  Validated   │  AI Service   │
  │  │  code         │                 │  Auth layer     │  request     │  (LLM, STT,   │
  │  │               │  ◄───────────   │                 │  ◄──────────  │   TTS, Vision)│
  │  │  • build_payload()  JSON resp   │  • Rate limits  │  Model resp  │               │
  │  │  • call_api()       + status    │  • Auth check   │  + tokens    │  • Inference   │
  │  │  • parse_response() code        │  • Logging      │  + usage     │  • Tokenize    │
  │  │  • log_result()                 │                 │              │  • Generate    │
  │  └──────────────┘                  └────────────────┘              └──────────────┘
  │                                                                          │
  │  This notebook implements Block A: everything your code does before,     │
  │  during, and after the HTTP call.                                        │
  └──────────────────────────────────────────────────────────────────────────┘
```

**Section overview:**

| Section | Implements | What it does |
|---------|------------|-------------|
| §2 Setup | Block A (config) | Load secrets, config, constants |
| §3 Script structure | Block A (design) | Modular functions for build/call/parse |
| §4 JSON, CSV, payloads | Block A (input) | Transform data into API request bodies |
| §5 requests library | Block A → B (send) | HTTP POST with auth headers |
| §6 Error handling | Block A ← B (receive) | Handle 4xx/5xx, timeouts, retries |
| §7 Logging | Block A (observe) | Trace every call with IDs, timing, tokens |
| §8 Production pipeline | Block A (end-to-end) | CSV → API → results CSV |
| §9 Frameworks | Block A (toolbox) | Libraries and their key functions |

### Production deployment context

```
  ┌───────────────┐     ┌───────────────┐     ┌────────────────┐     ┌──────────────┐
  │ Your app /    │────►│ Load balancer  │────►│ API gateway    │────►│ AI service   │
  │ cron job /    │     │ (nginx, ALB)   │     │ (rate limit,   │     │ (OpenAI,     │
  │ internal tool │◄────│                │◄────│  auth, log)    │◄────│  Azure, HF)  │
  └───────────────┘     └───────────────┘     └────────────────┘     └──────────────┘
        ▲ Block A             Network              Block B              Block C
        │
  Config (.env, config.json)
  Logs (structured JSON → log aggregation)
  Results (CSV, database, message queue)
```

The patterns in this notebook — config separation, retry logic, correlation IDs, structured logging — are exactly what you would use whether this code runs as a script, a cron job, a FastAPI endpoint, or a Celery task.

## 2. Setup — load secrets and config (run this cell first)

**Production pattern:** Secrets (API keys) in `.env`, never in code. Non-secret settings (API URLs, model names, timeouts) in `config.json`. Both loaded once at startup; every function reads from `CONFIG` and `os.environ`.

In [67]:
import os
import json
from pathlib import Path
from dotenv import load_dotenv

# Load secrets from .env (copy .env.example → .env and set OPENAI_API_KEY)
load_dotenv(Path.cwd() / ".env")

# Load non-secret config (API base URL, model names, timeouts)
with open(Path.cwd() / "config.json", encoding="utf-8") as f:
    CONFIG = json.load(f)

API_KEY = os.environ.get("OPENAI_API_KEY", "")
API_BASE = CONFIG.get("openai_api_base", "https://api.openai.com/v1").rstrip("/")
MODEL_CHAT = CONFIG.get("model_chat", "gpt-4o-mini")
TIMEOUT = CONFIG.get("request_timeout_seconds", 60)

print("Config loaded")
print(f"  API base:  {API_BASE}")
print(f"  Model:     {MODEL_CHAT}")
print(f"  Timeout:   {TIMEOUT}s")
print(f"  API key:   {'set (' + API_KEY[:8] + '...)' if API_KEY else 'NOT SET — add to .env'}")

Config loaded
  API base:  https://api.openai.com/v1
  Model:     gpt-4o-mini
  Timeout:   60s
  API key:   set (sk-proj-...)


In [68]:
# Environment validation — run this anytime to check your setup
def validate_environment() -> dict:
    """Check that all required config and secrets are present. Returns status dict."""
    checks = {
        "python_version": f"{__import__('sys').version}",
        "api_key_set": bool(API_KEY),
        "api_base": API_BASE,
        "model_chat": MODEL_CHAT,
        "config_json_loaded": bool(CONFIG),
        "tickets_csv_exists": Path("tickets.csv").exists(),
    }
    for k, v in checks.items():
        status = "OK" if v else "MISSING"
        print(f"  {k}: {v} [{status}]")
    return checks

env = validate_environment()

  python_version: 3.11.8 (tags/v3.11.8:db85d51, Feb  6 2024, 22:03:32) [MSC v.1937 64 bit (AMD64)] [OK]
  api_key_set: True [OK]
  api_base: https://api.openai.com/v1 [OK]
  model_chat: gpt-4o-mini [OK]
  config_json_loaded: True [OK]
  tickets_csv_exists: True [OK]


## 3. Structuring Python scripts for AI projects

**Real-life scenario:** You are building an internal tool that calls a summarization API for long reports. Leadership receives 10+ page documents weekly and needs 3-bullet executive summaries. Your code must be modular (easy to test, extend, and hand off to another developer), config-driven (no hardcoded URLs or models), and resilient (one bad report does not crash the entire batch).

**Design principles:**
- One function per responsibility: `build_payload`, `call_api`, `parse_response`
- Config and secrets loaded once at the top; functions receive them as parameters or read globals
- Type hints for every function signature (Python 3.11: `str | None`, `list[dict]`)
- Docstrings that state purpose, inputs, and outputs

### Project structure convention

For a production AI project, organize files by responsibility:

```
  summarization-service/
  ├── config/
  │   ├── config.json          # API URLs, model names, timeouts
  │   └── .env                 # Secrets (OPENAI_API_KEY)
  ├── services/
  │   ├── payload_builder.py   # build_payload(), build_batch_payload()
  │   ├── api_client.py        # call_api(), call_with_retry()
  │   └── response_parser.py   # parse_response(), extract_content()
  ├── utils/
  │   ├── logging_setup.py     # configure_logging(), get_logger()
  │   └── csv_handler.py       # read_tickets(), write_results()
  ├── main.py                  # Orchestrate: read → build → call → parse → write
  ├── requirements.txt
  └── tests/
      └── test_payload.py
```

In this notebook, we keep everything in cells for teaching. In production, each function below would live in its own module.

In [69]:
import requests

# --- Payload builder (would be services/payload_builder.py in production) ---

def build_chat_payload(
    system_content: str,
    user_content: str,
    model: str = MODEL_CHAT,
    max_tokens: int = 300,
    temperature: float = 0.3,
) -> dict:
    """Build OpenAI-style chat/completions request body.
    Truncate user content to 8000 chars to stay within context window.
    """
    return {
        "model": model,
        "messages": [
            {"role": "system", "content": system_content},
            {"role": "user", "content": user_content[:8000]},
        ],
        "max_tokens": max_tokens,
        "temperature": temperature,
    }


# --- API caller (would be services/api_client.py) ---

def call_chat_api(
    payload: dict,
    api_base: str = API_BASE,
    api_key: str = "",
    timeout: int = TIMEOUT,
) -> requests.Response:
    """POST payload to chat/completions endpoint. Returns raw Response object."""
    api_key = api_key or API_KEY
    url = f"{api_base}/chat/completions"
    headers = {"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"}
    resp = requests.post(url, json=payload, headers=headers, timeout=timeout)
    resp.raise_for_status()
    return resp


# --- Response parser (would be services/response_parser.py) ---

def parse_chat_response(resp_json: dict) -> tuple[str, dict]:
    """Extract content and metadata from chat/completions response.
    Returns (content_text, usage_metadata).
    """
    choices = resp_json.get("choices") or [{}]
    choice = choices[0]
    content = choice.get("message", {}).get("content", "")
    usage = resp_json.get("usage", {})
    meta = {
        "model": resp_json.get("model"),
        "finish_reason": choice.get("finish_reason"),
        "prompt_tokens": usage.get("prompt_tokens"),
        "completion_tokens": usage.get("completion_tokens"),
        "total_tokens": usage.get("total_tokens"),
    }
    return content, meta


print("Three core functions defined: build_chat_payload, call_chat_api, parse_chat_response")
print("Each maps to one step: build request (Block A) -> send (A->B) -> parse (B->A)")

Three core functions defined: build_chat_payload, call_chat_api, parse_chat_response
Each maps to one step: build request (Block A) -> send (A->B) -> parse (B->A)


In [70]:
# Demonstrate the build → call → parse flow (Scenario: executive summary)

sample_report = """
Q3 results show revenue up 12% YoY driven by enterprise adoption. Operations
expanded to two new regions (APAC and LATAM). Customer satisfaction scores
improved from 4.1 to 4.5. Two product launches (Analytics Pro and Security Suite)
delayed to Q4 due to compliance review. Headcount grew by 8% with 15 new hires
in engineering. Cloud infrastructure costs reduced by 18% after migration to
reserved instances. Three enterprise contracts signed worth $2.4M ARR combined.
Churn rate decreased from 5.2% to 3.8%. Net promoter score reached 62.
"""

# Step 1: Build payload
payload = build_chat_payload(
    system_content="You are an assistant that writes concise executive summaries. Output exactly 3 bullet points.",
    user_content=sample_report,
    max_tokens=300,
)
print("Request payload (first 400 chars):")
print(json.dumps(payload, indent=2)[:400])

Request payload (first 400 chars):
{
  "model": "gpt-4o-mini",
  "messages": [
    {
      "role": "system",
      "content": "You are an assistant that writes concise executive summaries. Output exactly 3 bullet points."
    },
    {
      "role": "user",
      "content": "\nQ3 results show revenue up 12% YoY driven by enterprise adoption. Operations\nexpanded to two new regions (APAC and LATAM). Customer satisfaction scores\nimpr


In [71]:
# Step 2 & 3: Call API and parse response

def summarize_for_executive(text: str) -> tuple[str | None, dict]:
    """Use LLM API to produce a 3-bullet executive summary.
    Returns (summary_text, usage_metadata).
    """
    if not API_KEY:
        return "[Set OPENAI_API_KEY in .env to run real API calls]", {}
    payload = build_chat_payload(
        "You are an assistant that writes concise executive summaries. Output exactly 3 bullet points.",
        text,
    )
    resp = call_chat_api(payload)
    content, meta = parse_chat_response(resp.json())
    return content, meta

summary, usage = summarize_for_executive(sample_report)
print("Executive summary:")
print(summary)
print("\nUsage metadata:", usage)

Executive summary:
- Q3 revenue increased by 12% YoY, bolstered by enterprise adoption and three new contracts totaling $2.4M ARR.
- Customer satisfaction improved significantly, with scores rising from 4.1 to 4.5 and churn rate decreasing from 5.2% to 3.8%.
- Operations expanded into APAC and LATAM, while cloud infrastructure costs were reduced by 18%; however, two product launches were delayed to Q4 due to compliance issues.

Usage metadata: {'model': 'gpt-4o-mini-2024-07-18', 'finish_reason': 'stop', 'prompt_tokens': 163, 'completion_tokens': 99, 'total_tokens': 262}


### Modular pipeline design pattern

**Real-life scenario:** Your team maintains a risk/compliance classifier. The same modular pattern (build → call → parse) is reused with only the system prompt changed. Functions compose into pipelines where each step is independently testable.

In [72]:
# Same build → call → parse pattern; only the system prompt changes

def classify_risk(text: str) -> tuple[str | None, dict]:
    """Classify text into risk levels: low, medium, high.
    Returns (classification_text, usage_metadata).
    """
    if not API_KEY:
        return "[Set OPENAI_API_KEY in .env]", {}
    payload = build_chat_payload(
        system_content=(
            "You are a risk analyst. Classify the following text as low, medium, or high risk. "
            "Output exactly one line: RISK: <level> — <one-sentence reason>."
        ),
        user_content=text,
        max_tokens=100,
        temperature=0.1,
    )
    resp = call_chat_api(payload)
    return parse_chat_response(resp.json())


def generate_kb_article(faq_question: str, answer_notes: str) -> tuple[str | None, dict]:
    """Generate a knowledge base article from a FAQ question and rough answer notes."""
    if not API_KEY:
        return "[Set OPENAI_API_KEY in .env]", {}
    payload = build_chat_payload(
        system_content=(
            "You write concise knowledge base articles. Structure: ## Problem, "
            "## Solution (numbered steps), ## Additional notes (if needed). "
            "Be precise. Do not add information not in the input."
        ),
        user_content=f"Question: {faq_question}\nAnswer notes: {answer_notes}",
        max_tokens=500,
    )
    resp = call_chat_api(payload)
    return parse_chat_response(resp.json())


# Example: risk classification
risk_text = "Employee shared customer PII via unencrypted email to external vendor."
risk_result, risk_meta = classify_risk(risk_text)
print("Risk classification:", risk_result)
print("Tokens used:", risk_meta.get("total_tokens"))

Risk classification: RISK: High — Sharing customer PII via unencrypted email poses a significant risk of data breach and non-compliance with privacy regulations.
Tokens used: 90


In [73]:
# Example: knowledge base article generation
kb_result, kb_meta = generate_kb_article(
    faq_question="How do I reset my password?",
    answer_notes="Go to login page, click Forgot Password, enter email, check inbox, click link, set new password. If link expired, repeat. If error 500, clear browser cache or try incognito.",
)
print("Generated KB article:")
print(kb_result)

Generated KB article:
## Problem
User needs to reset their password.

## Solution
1. Go to the login page.
2. Click on "Forgot Password."
3. Enter your email address.
4. Check your inbox for the password reset email.
5. Click the link in the email.
6. Set a new password.

## Additional notes
- If the link has expired, repeat the process.
- If you encounter a 500 error, clear your browser cache or try using incognito mode.


### Type-safe configuration with dataclasses

**Production pattern:** Instead of accessing `CONFIG["model_chat"]` everywhere (typo-prone, no autocompletion), wrap config in a dataclass. This gives you type checking, default values, and IDE support.

In [74]:
from dataclasses import dataclass

@dataclass(frozen=True)
class AIServiceConfig:
    """Typed configuration for AI API calls. Loaded from config.json + .env."""
    api_base: str
    api_key: str
    model_chat: str
    timeout: int
    max_tokens_default: int = 300
    temperature_default: float = 0.3

    @classmethod
    def from_env(cls) -> "AIServiceConfig":
        """Factory method: build config from environment and config.json."""
        return cls(
            api_base=CONFIG.get("openai_api_base", "https://api.openai.com/v1").rstrip("/"),
            api_key=os.environ.get("OPENAI_API_KEY", ""),
            model_chat=CONFIG.get("model_chat", "gpt-4o-mini"),
            timeout=CONFIG.get("request_timeout_seconds", 60),
        )

cfg = AIServiceConfig.from_env()
print(f"Config: model={cfg.model_chat}, base={cfg.api_base}, timeout={cfg.timeout}s")
print(f"API key set: {bool(cfg.api_key)}")

Config: model=gpt-4o-mini, base=https://api.openai.com/v1, timeout=60s
API key set: True


## 4. Working with JSON, CSV, and API payloads

**Real-life scenario:** Your operations team uploads 50 support tickets weekly as CSV. Your tool reads the CSV, builds API payloads for classification, sends them, parses responses, and writes results back to a new CSV — all without manual intervention.

### Reading CSV with DictReader

**Why DictReader:** Returns each row as a dict (column name → value) instead of a list. You access `row["subject"]` instead of `row[1]`. Safer, readable, and the column names match your API payload field names.

In [75]:
import csv

def read_tickets(csv_path: str = "tickets.csv") -> list[dict]:
    """Read tickets from CSV file. Returns list of dicts with keys: ticket_id, subject, body, priority."""
    tickets = []
    with open(csv_path, encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            tickets.append(row)
    return tickets

tickets = read_tickets()
print(f"Loaded {len(tickets)} tickets from tickets.csv")
print(f"Columns: {list(tickets[0].keys())}")
print(f"\nFirst 3 tickets:")
for t in tickets[:3]:
    print(f"  #{t['ticket_id']} [{t['priority']}] {t['subject']}: {t['body'][:60]}...")

Loaded 18 tickets from tickets.csv
Columns: ['ticket_id', 'subject', 'body', 'priority']

First 3 tickets:
  #1 [high] Login failed: Getting error 500 when I click Forgot password. Tried from C...
  #2 [high] Invoice discrepancy: My last invoice shows double charge for the premium plan. Ca...
  #3 [medium] Export to PDF: We need an option to export reports to PDF for our monthly b...


In [ ]:
# Inspect full data shape and priority distribution
priorities = {}
for t in tickets:
    p = t.get("priority", "unknown")
    priorities[p] = priorities.get(p, 0) + 1

print("Ticket distribution by priority:")
for priority, count in sorted(priorities.items()):
    print(f"  {priority}: {count}")
print(f"\nAverage body length: {sum(len(t['body']) for t in tickets) / len(tickets):.0f} chars")

Ticket distribution by priority:
  high: 8
  low: 3
  medium: 7

Average body length: 77 chars


### Building API payloads from CSV rows

**Pattern:** For each ticket, build a chat/completions payload where the system prompt defines the classification task and the user message contains the ticket text.

In [77]:
CLASSIFY_SYSTEM_PROMPT = (
    "You classify support tickets into exactly one category. "
    "Categories: billing, technical, feature_request, account, security. "
    "Output exactly one line: CATEGORY: <category>"
)

def build_classify_payload(ticket: dict) -> dict:
    """Build classification payload for a single ticket."""
    user_content = f"Subject: {ticket['subject']}\nBody: {ticket['body']}"
    return build_chat_payload(
        system_content=CLASSIFY_SYSTEM_PROMPT,
        user_content=user_content,
        max_tokens=50,
        temperature=0.0,
    )

# Show payload for first ticket
sample_payload = build_classify_payload(tickets[0])
print("Classification payload for ticket #1:")
print(json.dumps(sample_payload, indent=2))

Classification payload for ticket #1:
{
  "model": "gpt-4o-mini",
  "messages": [
    {
      "role": "system",
      "content": "You classify support tickets into exactly one category. Categories: billing, technical, feature_request, account, security. Output exactly one line: CATEGORY: <category>"
    },
    {
      "role": "user",
      "content": "Subject: Login failed\nBody: Getting error 500 when I click Forgot password. Tried from Chrome and mobile."
    }
  ],
  "max_tokens": 50,
  "temperature": 0.0
}


In [78]:
# Build payloads for all tickets (batch preparation)
all_payloads = [build_classify_payload(t) for t in tickets]
print(f"Built {len(all_payloads)} payloads")
print(f"Payload keys: {list(all_payloads[0].keys())}")
print(f"Model: {all_payloads[0]['model']}")
print(f"Max tokens per call: {all_payloads[0]['max_tokens']}")

Built 18 payloads
Payload keys: ['model', 'messages', 'max_tokens', 'temperature']
Model: gpt-4o-mini
Max tokens per call: 50


### Classifying tickets one by one (single-call pattern)

**Production pattern:** Call the API for each ticket. Handle failures per ticket so one bad response does not crash the batch.

In [79]:
def classify_ticket(ticket: dict) -> dict:
    """Classify a single ticket. Returns ticket dict with 'category' added."""
    if not API_KEY:
        return {**ticket, "category": "[no_api_key]"}
    payload = build_classify_payload(ticket)
    try:
        resp = call_chat_api(payload)
        content, meta = parse_chat_response(resp.json())
        # Parse "CATEGORY: billing" → "billing"
        if "CATEGORY:" in content.upper():
            category = content.split(":", 1)[1].strip().lower()
        else:
            category = content.strip().lower()
        return {**ticket, "category": category, "tokens": meta.get("total_tokens")}
    except Exception as e:
        return {**ticket, "category": f"[error: {e}]"}

# Classify first 3 tickets as demo
print("Classifying first 3 tickets:")
for t in tickets[:3]:
    result = classify_ticket(t)
    print(f"  #{result['ticket_id']} {result['subject']} → {result['category']}")

Classifying first 3 tickets:
  #1 Login failed → technical
  #2 Invoice discrepancy → billing
  #3 Export to PDF → feature_request


### Batch classification with skip-failures

**Production pattern:** Process all tickets in a loop. If one fails (timeout, bad response, rate limit), log the error and continue. Return results for successful calls and error markers for failures.

In [80]:
def classify_batch(tickets: list[dict], skip_failures: bool = True) -> list[dict]:
    """Classify a batch of tickets. Returns list of tickets with 'category' added.
    If skip_failures=True, errors are logged but processing continues.
    """
    results = []
    success_count = 0
    error_count = 0
    for i, ticket in enumerate(tickets):
        result = classify_ticket(ticket)
        if result["category"].startswith("[error"):
            error_count += 1
            if not skip_failures:
                raise RuntimeError(f"Failed on ticket #{ticket['ticket_id']}: {result['category']}")
        else:
            success_count += 1
        results.append(result)
    print(f"Batch complete: {success_count} success, {error_count} errors out of {len(tickets)} tickets")
    return results

# Run on first 5 tickets
batch_results = classify_batch(tickets[:5])
for r in batch_results:
    print(f"  #{r['ticket_id']} [{r.get('priority', '?')}] → {r['category']}")

Batch complete: 5 success, 0 errors out of 5 tickets
  #1 [high] → technical
  #2 [high] → billing
  #3 [medium] → feature_request
  #4 [high] → technical
  #5 [low] → billing


### Writing results back to CSV (DictWriter)

**Production pattern:** After classification, write augmented data to a new CSV. Operations team can review, override, and import into their ticketing system.

In [81]:
def write_results(results: list[dict], output_path: str = "tickets_classified.csv") -> str:
    """Write classified tickets to CSV. Returns output path."""
    if not results:
        return output_path
    fieldnames = list(results[0].keys())
    with open(output_path, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(results)
    print(f"Wrote {len(results)} rows to {output_path}")
    return output_path

# Write results from the batch above
write_results(batch_results)

Wrote 5 rows to tickets_classified.csv


'tickets_classified.csv'

### Nested JSON response parsing

**Production pattern:** API responses contain nested structures (choices, usage, metadata). Parse defensively — missing keys should return defaults, not crash your pipeline.

In [82]:
# Deep parse example: extract all useful fields from a chat/completions response

sample_api_response = {
    "id": "chatcmpl-abc123",
    "object": "chat.completion",
    "created": 1710000000,
    "model": "gpt-4o-mini",
    "choices": [
        {
            "index": 0,
            "message": {"role": "assistant", "content": "CATEGORY: technical"},
            "finish_reason": "stop",
        }
    ],
    "usage": {"prompt_tokens": 42, "completion_tokens": 5, "total_tokens": 47},
}

def deep_parse_response(resp_json: dict) -> dict:
    """Extract all useful fields from API response into a flat dict."""
    choices = resp_json.get("choices") or [{}]
    choice = choices[0]
    usage = resp_json.get("usage", {})
    return {
        "response_id": resp_json.get("id"),
        "model": resp_json.get("model"),
        "content": choice.get("message", {}).get("content", ""),
        "finish_reason": choice.get("finish_reason"),
        "prompt_tokens": usage.get("prompt_tokens", 0),
        "completion_tokens": usage.get("completion_tokens", 0),
        "total_tokens": usage.get("total_tokens", 0),
        "created": resp_json.get("created"),
    }

parsed = deep_parse_response(sample_api_response)
print("Parsed response (flat dict):")
for k, v in parsed.items():
    print(f"  {k}: {v}")

Parsed response (flat dict):
  response_id: chatcmpl-abc123
  model: gpt-4o-mini
  content: CATEGORY: technical
  finish_reason: stop
  prompt_tokens: 42
  completion_tokens: 5
  total_tokens: 47
  created: 1710000000


In [83]:
# Defensive parsing: handle missing or malformed responses

malformed_responses = [
    {},                                          # Empty response
    {"choices": []},                              # Empty choices list
    {"choices": [{"message": {}}]},               # No content key
    {"choices": [{"message": {"content": None}}]}, # Null content
]

print("Defensive parsing of malformed responses:")
for i, resp in enumerate(malformed_responses):
    parsed = deep_parse_response(resp)
    print(f"  Response {i}: content='{parsed['content']}', tokens={parsed['total_tokens']}")

Defensive parsing of malformed responses:
  Response 0: content='', tokens=0
  Response 1: content='', tokens=0
  Response 2: content='', tokens=0
  Response 3: content='None', tokens=0


### JSON payload validation

**Production pattern:** Before sending a payload to the API, validate that required fields are present and of correct types. Catch errors before they hit the network.

In [84]:
def validate_chat_payload(payload: dict) -> list[str]:
    """Validate chat/completions payload. Returns list of error messages (empty if valid)."""
    errors = []
    if "model" not in payload:
        errors.append("Missing 'model'")
    if "messages" not in payload:
        errors.append("Missing 'messages'")
    elif not isinstance(payload["messages"], list):
        errors.append("'messages' must be a list")
    elif len(payload["messages"]) == 0:
        errors.append("'messages' must not be empty")
    else:
        for i, msg in enumerate(payload["messages"]):
            if "role" not in msg:
                errors.append(f"Message {i}: missing 'role'")
            if "content" not in msg:
                errors.append(f"Message {i}: missing 'content'")
    if "max_tokens" in payload and not isinstance(payload["max_tokens"], int):
        errors.append("'max_tokens' must be an integer")
    return errors

# Valid payload
valid = build_chat_payload("Summarize.", "Some text here.")
print("Valid payload errors:", validate_chat_payload(valid))

# Invalid payload
invalid = {"messages": [{"role": "user"}]}  # missing model, missing content
print("Invalid payload errors:", validate_chat_payload(invalid))

Valid payload errors: []
Invalid payload errors: ["Missing 'model'", "Message 0: missing 'content'"]


## 5. Using the `requests` library for API integration

**REST fundamentals:**

| Concept | What it is | Example in this notebook |
|---------|-----------|-------------------------|
| Endpoint | URL path for a resource | `/v1/chat/completions` |
| Method | HTTP verb (GET, POST, PUT, DELETE) | `POST` (send data, get result) |
| Headers | Metadata (auth, content type) | `Authorization: Bearer sk-...` |
| Body | Request payload (JSON for POST) | `{"model": "...", "messages": [...]}` |
| Status code | Response result code | `200 OK`, `429 Rate limited`, `500 Server error` |
| Response body | Result data (JSON) | `{"choices": [...], "usage": {...}}` |

**Real-life scenario:** You call a chat/completions-style endpoint to generate support reply drafts. The same `requests.post()` pattern is used for every AI API — only the URL, headers, and body change.

In [85]:
# Basic POST request — the fundamental pattern for all AI API calls

def call_api_basic(messages: list[dict], model: str = MODEL_CHAT, max_tokens: int = 300) -> dict | None:
    """Minimal chat/completions call. Returns response JSON or None."""
    if not API_KEY:
        print("[Set OPENAI_API_KEY in .env]")
        return None
    url = f"{API_BASE}/chat/completions"
    headers = {
        "Authorization": f"Bearer {API_KEY}",  # Authentication
        "Content-Type": "application/json",     # Tell server: body is JSON
    }
    body = {
        "model": model,
        "messages": messages,
        "max_tokens": max_tokens,
    }
    # POST = send data and get result (vs GET = read-only)
    resp = requests.post(url, json=body, headers=headers, timeout=TIMEOUT)
    resp.raise_for_status()  # Raise exception for 4xx/5xx
    return resp.json()

# Example: simple question
resp_data = call_api_basic([
    {"role": "user", "content": "What are three best practices for API security? Be concise."}
])
if resp_data:
    print(resp_data["choices"][0]["message"]["content"])

1. **Authentication and Authorization**: Implement strong authentication mechanisms (e.g., OAuth, API keys) and ensure proper authorization to control access to API resources.

2. **Input Validation and Sanitization**: Rigorously validate and sanitize all incoming data to prevent injection attacks and ensure that only expected and safe inputs are processed.

3. **Rate Limiting and Throttling**: Enforce rate limits to mitigate abuse and denial-of-service attacks by controlling the number of requests a user can make within a specific timeframe.


### Common API parameters

| Parameter | Type | What it controls | Typical value |
|-----------|------|-----------------|---------------|
| `model` | string | Which model to use | `"gpt-4o-mini"`, `"gpt-4o"` |
| `messages` | list[dict] | Conversation history | `[{"role": "system", ...}, {"role": "user", ...}]` |
| `max_tokens` | int | Max response length (tokens) | 100–4000 |
| `temperature` | float | Randomness (0=deterministic, 2=creative) | 0.0–0.7 for factual tasks |
| `top_p` | float | Nucleus sampling threshold | 0.9–1.0 |
| `response_format` | dict | Force output format | `{"type": "json_object"}` |
| `stream` | bool | Stream response tokens | `false` for batch; `true` for real-time UI |

In [86]:
# Support reply draft — system + user messages (Block A: build messages)

def build_support_messages(ticket_text: str, knowledge_snippet: str) -> list[dict]:
    """Construct messages for support reply generation. System sets behavior; user provides context."""
    return [
        {
            "role": "system",
            "content": (
                "You write brief, professional support replies. Use only the provided knowledge. "
                "Be concise and actionable. Do not invent information."
            ),
        },
        {
            "role": "user",
            "content": f"Knowledge:\n{knowledge_snippet}\n\nCustomer ticket:\n{ticket_text}",
        },
    ]

ticket_text = "Getting error 500 when I click Forgot password. Tried from Chrome and mobile."
knowledge = (
    "Password reset: use the Forgot password link on login page. "
    "If error 500: clear browser cache or try incognito mode. "
    "If still failing: contact support with browser version and screenshot."
)
messages = build_support_messages(ticket_text, knowledge)
print("Messages structure:")
for m in messages:
    print(f"  [{m['role']}] {m['content'][:80]}...")

Messages structure:
  [system] You write brief, professional support replies. Use only the provided knowledge. ...
  [user] Knowledge:
Password reset: use the Forgot password link on login page. If error ...


In [87]:
# Call API with support messages
def draft_support_reply(ticket_text: str, knowledge_snippet: str) -> str | None:
    """Generate support reply draft. Human-in-the-loop: review before sending to customer."""
    if not API_KEY:
        return "[Set OPENAI_API_KEY in .env]"
    messages = build_support_messages(ticket_text, knowledge_snippet)
    resp_data = call_api_basic(messages, max_tokens=200)
    if resp_data:
        return resp_data["choices"][0]["message"]["content"]
    return None

reply = draft_support_reply(ticket_text, knowledge)
print("Draft support reply (human reviews before sending):")
print(reply)

Draft support reply (human reviews before sending):
Please try clearing your browser cache or use incognito mode. If the issue persists, contact support again and include your browser version and a screenshot of the error.


### Request and response inspection

**Production debugging:** When an API call behaves unexpectedly, inspect the raw request and response to understand what was sent and what came back.

In [88]:
def call_api_with_inspection(
    messages: list[dict], model: str = MODEL_CHAT, max_tokens: int = 200
) -> dict:
    """Call API and return both response data and inspection metadata."""
    if not API_KEY:
        return {"content": "[no key]", "inspection": {}}
    url = f"{API_BASE}/chat/completions"
    headers = {"Authorization": f"Bearer {API_KEY}", "Content-Type": "application/json"}
    body = {"model": model, "messages": messages, "max_tokens": max_tokens}
    resp = requests.post(url, json=body, headers=headers, timeout=TIMEOUT)
    resp.raise_for_status()
    data = resp.json()
    content, meta = parse_chat_response(data)
    inspection = {
        "status_code": resp.status_code,
        "elapsed_ms": resp.elapsed.total_seconds() * 1000,
        "response_headers": dict(resp.headers),
        "request_url": resp.request.url,
        "request_method": resp.request.method,
        "content_length": len(resp.content),
    }
    return {"content": content, "meta": meta, "inspection": inspection}

result = call_api_with_inspection([{"role": "user", "content": "Say hello in one word."}])
print("Content:", result["content"])
print(f"\nInspection:")
print(f"  Status: {result['inspection'].get('status_code')}")
print(f"  Elapsed: {result['inspection'].get('elapsed_ms', 0):.0f}ms")
print(f"  Response size: {result['inspection'].get('content_length', 0)} bytes")
print(f"  URL: {result['inspection'].get('request_url')}")

Content: Hello!

Inspection:
  Status: 200
  Elapsed: 981ms
  Response size: 812 bytes
  URL: https://api.openai.com/v1/chat/completions


### Session-based requests (connection pooling)

**Real-life scenario:** Your service makes 200+ API calls per hour. Creating a new TCP connection for each call adds latency. `requests.Session()` reuses connections (HTTP keep-alive) and lets you set default headers once.

```
  Without Session: new TCP connection per call → ~100ms overhead each
  With Session:    reuse connections           → ~10ms overhead each
```

In [89]:
import time

def create_api_session(api_key: str = "", timeout: int = TIMEOUT) -> requests.Session:
    """Create a session with default headers and timeout for reuse across multiple calls."""
    api_key = api_key or API_KEY
    session = requests.Session()
    session.headers.update({
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
    })
    return session


def call_with_session(
    session: requests.Session, messages: list[dict], model: str = MODEL_CHAT, max_tokens: int = 200
) -> tuple[str | None, float]:
    """Call API using session (connection pooling). Returns (content, elapsed_ms)."""
    url = f"{API_BASE}/chat/completions"
    body = {"model": model, "messages": messages, "max_tokens": max_tokens}
    start = time.monotonic()
    resp = session.post(url, json=body, timeout=TIMEOUT)
    elapsed_ms = (time.monotonic() - start) * 1000
    resp.raise_for_status()
    content = resp.json()["choices"][0]["message"]["content"]
    return content, elapsed_ms


# Example: 3 sequential calls using the same session
if API_KEY:
    session = create_api_session()
    questions = [
        "Name one benefit of connection pooling.",
        "What HTTP header carries the API key?",
        "What status code means rate limited?",
    ]
    for q in questions:
        content, ms = call_with_session(session, [{"role": "user", "content": q + " One sentence."}])
        print(f"  [{ms:.0f}ms] {q} → {content[:80]}")
    session.close()
else:
    print("Session example: create_api_session() reuses TCP connections across calls.")

  [1266ms] Name one benefit of connection pooling. → Connection pooling improves application performance by reusing existing database
  s] What HTTP header carries the API key? → The API key is typically sent in the "Authorization" HTTP header.
  s] What status code means rate limited? → The status code that typically indicates a rate limit has been exceeded is 429 T


### Timeout configuration

**Production pattern:** Use a tuple `(connect_timeout, read_timeout)` instead of a single number. Connect timeout catches DNS/network issues fast. Read timeout gives the AI model time to generate a response.

In [90]:
# Timeout tuple: (connect_timeout_seconds, read_timeout_seconds)
# connect: how long to wait to establish connection (network layer)
# read: how long to wait for the response body (model inference time)

TIMEOUT_TUPLE = (5, 60)  # 5s to connect, 60s to read

def call_api_with_timeout(
    messages: list[dict],
    timeout: tuple[int, int] = TIMEOUT_TUPLE,
) -> dict | None:
    """Call API with separate connect and read timeouts."""
    if not API_KEY:
        return None
    url = f"{API_BASE}/chat/completions"
    headers = {"Authorization": f"Bearer {API_KEY}", "Content-Type": "application/json"}
    body = {"model": MODEL_CHAT, "messages": messages, "max_tokens": 200}
    try:
        resp = requests.post(url, json=body, headers=headers, timeout=timeout)
        resp.raise_for_status()
        return resp.json()
    except requests.exceptions.ConnectTimeout:
        print(f"Connect timeout after {timeout[0]}s — check network or DNS")
        return None
    except requests.exceptions.ReadTimeout:
        print(f"Read timeout after {timeout[1]}s — model may be overloaded")
        return None

print(f"Timeout config: connect={TIMEOUT_TUPLE[0]}s, read={TIMEOUT_TUPLE[1]}s")
print("Use tuple timeouts in production to distinguish network from inference delays.")

Timeout config: connect=5s, read=60s
Use tuple timeouts in production to distinguish network from inference delays.


### Authentication patterns

| Pattern | Where key goes | Example | When used |
|---------|---------------|---------|----------|
| Bearer token | `Authorization` header | `Bearer sk-abc...` | OpenAI, Azure OpenAI, most SaaS APIs |
| API key header | Custom header | `X-API-Key: abc...` | Some internal/enterprise APIs |
| API key in query | URL parameter | `?api_key=abc...` | Legacy APIs (less secure, logged in URLs) |
| OAuth2 client creds | Token exchange → Bearer | Exchange client_id + secret → access_token | Enterprise SSO, Google APIs |

In [91]:
# Three authentication patterns side by side

# 1. Bearer token (used in this course)
headers_bearer = {
    "Authorization": f"Bearer {API_KEY}",
    "Content-Type": "application/json",
}

# 2. API key in custom header (some enterprise APIs)
headers_custom = {
    "X-API-Key": API_KEY,
    "Content-Type": "application/json",
}

# 3. API key as query parameter (legacy pattern; avoid — key appears in logs)
# requests.get(url, params={"api_key": API_KEY})

print("Authentication patterns:")
print(f"  Bearer: Authorization: Bearer {API_KEY[:8]}..." if API_KEY else "  Bearer: Authorization: Bearer <not set>")
print(f"  Custom header: X-API-Key: {API_KEY[:8]}..." if API_KEY else "  Custom header: X-API-Key: <not set>")
print("  Query param: ?api_key=... (avoid in production — key in server logs)")

Authentication patterns:
  Bearer: Authorization: Bearer sk-proj-...
  Custom header: X-API-Key: sk-proj-...
  Query param: ?api_key=... (avoid in production — key in server logs)


## 6. Handling API responses and errors

**Real-life scenario:** Your production pipeline processes 1,000 documents per night. The AI API occasionally returns 429 (rate limit), times out under load, or returns 5xx during maintenance windows. Three failures must not halt the entire batch. You need:
- Retry with exponential backoff for transient errors (429, 5xx)
- Structured error parsing for actionable diagnostics
- Graceful degradation when the API is unreachable

### Error classification

| Status code | Meaning | Action |
|-------------|---------|--------|
| 200 | Success | Parse response |
| 400 | Bad request (invalid payload) | Fix payload; do not retry |
| 401 | Unauthorized (bad API key) | Check .env; do not retry |
| 403 | Forbidden (no access to model) | Check permissions; do not retry |
| 404 | Endpoint not found | Check URL; do not retry |
| 422 | Unprocessable (e.g., bad model name) | Fix parameters; do not retry |
| 429 | Rate limited | Retry after backoff |
| 500 | Server error | Retry with backoff |
| 502/503 | Gateway/service unavailable | Retry with backoff |
| Timeout | No response in time | Retry with longer timeout |

In [92]:
# Basic error handling: try/except with status code inspection

def safe_call(messages: list[dict], max_tokens: int = 200) -> tuple[str | None, str | None]:
    """Safe API call. Returns (content, error_message). One of them is None."""
    if not API_KEY:
        return None, "API key not set"
    url = f"{API_BASE}/chat/completions"
    headers = {"Authorization": f"Bearer {API_KEY}", "Content-Type": "application/json"}
    body = {"model": MODEL_CHAT, "messages": messages, "max_tokens": max_tokens}
    try:
        resp = requests.post(url, json=body, headers=headers, timeout=TIMEOUT)
        resp.raise_for_status()
        content = resp.json()["choices"][0]["message"]["content"]
        return content, None
    except requests.exceptions.Timeout:
        return None, f"Timeout after {TIMEOUT}s"
    except requests.exceptions.HTTPError as e:
        status = e.response.status_code if e.response is not None else "unknown"
        body_text = e.response.text[:200] if e.response is not None else ""
        return None, f"HTTP {status}: {body_text}"
    except requests.exceptions.RequestException as e:
        return None, f"Request failed: {e}"

# Example
content, error = safe_call([{"role": "user", "content": "Say OK"}])
if content:
    print("Success:", content)
else:
    print("Error:", error)

Success: OK!


### Structured error body parsing

**Production pattern:** API error responses typically include structured JSON with `error.message`, `error.type`, and `error.code`. Parse these for actionable diagnostics instead of just logging the status code.

In [93]:
def parse_api_error(response: requests.Response) -> dict:
    """Parse structured error from API response body."""
    error_info = {
        "status_code": response.status_code,
        "error_type": None,
        "error_message": None,
        "error_code": None,
        "raw_body": response.text[:500],
    }
    try:
        body = response.json()
        err = body.get("error", {})
        error_info["error_type"] = err.get("type")
        error_info["error_message"] = err.get("message")
        error_info["error_code"] = err.get("code")
    except (ValueError, KeyError):
        pass
    return error_info

# Simulate parsing a structured API error
class FakeErrorResponse:
    status_code = 429
    text = '{"error": {"message": "Rate limit exceeded. Retry after 2s.", "type": "rate_limit_error", "code": "rate_limit"}}'
    def json(self):
        return json.loads(self.text)

parsed_error = parse_api_error(FakeErrorResponse())
print("Parsed API error:")
for k, v in parsed_error.items():
    print(f"  {k}: {v}")

Parsed API error:
  status_code: 429
  error_type: rate_limit_error
  error_message: Rate limit exceeded. Retry after 2s.
  error_code: rate_limit
  raw_body: {"error": {"message": "Rate limit exceeded. Retry after 2s.", "type": "rate_limit_error", "code": "rate_limit"}}


### Exponential backoff with jitter

**Why jitter:** If 10 clients hit a rate limit at the same time and all retry after exactly 2 seconds, they all hit the limit again simultaneously. Adding random jitter spreads retries over time.

```
  Attempt 1: wait 1s + random(0, 0.5)s
  Attempt 2: wait 2s + random(0, 1.0)s
  Attempt 3: wait 4s + random(0, 2.0)s
  Formula:   wait = base_delay * 2^attempt + random(0, base_delay * 2^attempt * jitter_factor)
```

In [94]:
import random

RETRYABLE_STATUS_CODES = {429, 500, 502, 503}

def call_with_retry(
    messages: list[dict],
    max_retries: int = 3,
    base_delay: float = 1.0,
    jitter_factor: float = 0.5,
    max_tokens: int = 200,
) -> tuple[str | None, dict]:
    """Call API with exponential backoff and jitter on retryable errors.
    Returns (content, metadata) where metadata includes retry info.
    """
    if not API_KEY:
        return "[Set OPENAI_API_KEY in .env]", {"retries": 0}
    url = f"{API_BASE}/chat/completions"
    headers = {"Authorization": f"Bearer {API_KEY}", "Content-Type": "application/json"}
    body = {"model": MODEL_CHAT, "messages": messages, "max_tokens": max_tokens}
    last_error = None

    for attempt in range(max_retries + 1):
        try:
            resp = requests.post(url, json=body, headers=headers, timeout=TIMEOUT)
            resp.raise_for_status()
            data = resp.json()
            content, meta = parse_chat_response(data)
            meta["attempts"] = attempt + 1
            return content, meta
        except requests.exceptions.HTTPError as e:
            status = e.response.status_code if e.response is not None else 0
            last_error = f"HTTP {status}"
            if status not in RETRYABLE_STATUS_CODES or attempt == max_retries:
                return None, {"error": last_error, "attempts": attempt + 1}
            delay = base_delay * (2 ** attempt) + random.uniform(0, base_delay * (2 ** attempt) * jitter_factor)
            print(f"  Retry {attempt + 1}/{max_retries}: {status} — waiting {delay:.1f}s")
            time.sleep(delay)
        except requests.exceptions.Timeout:
            last_error = "Timeout"
            if attempt == max_retries:
                return None, {"error": "Timeout after all retries", "attempts": attempt + 1}
            delay = base_delay * (2 ** attempt)
            print(f"  Retry {attempt + 1}/{max_retries}: timeout — waiting {delay:.1f}s")
            time.sleep(delay)
        except requests.exceptions.RequestException as e:
            return None, {"error": str(e), "attempts": attempt + 1}

    return None, {"error": last_error, "attempts": max_retries + 1}

# Example
content, meta = call_with_retry([{"role": "user", "content": "Say OK."}])
print(f"Result: {content}")
print(f"Metadata: {meta}")

Result: OK.
Metadata: {'model': 'gpt-4o-mini-2024-07-18', 'finish_reason': 'stop', 'prompt_tokens': 10, 'completion_tokens': 2, 'total_tokens': 12, 'attempts': 1}


### Circuit breaker pattern

**Production pattern:** When an API is consistently failing (e.g., during maintenance), stop sending requests to avoid wasting time and hitting rate limits. A circuit breaker tracks failures and "opens" the circuit after a threshold, returning a default response immediately.

```
  CLOSED (normal)  ──failures >= threshold──►  OPEN (reject all)
       ▲                                          │
       │                                   after cooldown
       │                                          ▼
       └──────── success ◄──────────────  HALF-OPEN (try one)
```

In [95]:
class CircuitBreaker:
    """Simple counter-based circuit breaker for API calls."""

    def __init__(self, failure_threshold: int = 5, cooldown_seconds: float = 30.0):
        self.failure_threshold = failure_threshold
        self.cooldown_seconds = cooldown_seconds
        self.failures = 0
        self.state = "closed"  # closed, open, half-open
        self.last_failure_time: float = 0

    def can_call(self) -> bool:
        """Check if a call should be attempted."""
        if self.state == "closed":
            return True
        if self.state == "open":
            if time.monotonic() - self.last_failure_time >= self.cooldown_seconds:
                self.state = "half-open"
                return True
            return False
        return True  # half-open: allow one try

    def record_success(self) -> None:
        self.failures = 0
        self.state = "closed"

    def record_failure(self) -> None:
        self.failures += 1
        self.last_failure_time = time.monotonic()
        if self.failures >= self.failure_threshold:
            self.state = "open"

    def __repr__(self) -> str:
        return f"CircuitBreaker(state={self.state}, failures={self.failures}/{self.failure_threshold})"


# Example usage
breaker = CircuitBreaker(failure_threshold=3, cooldown_seconds=10)
print(f"Initial: {breaker}")

# Simulate 3 failures
for i in range(3):
    breaker.record_failure()
    print(f"After failure {i+1}: {breaker}")

print(f"Can call? {breaker.can_call()}")

# After success
breaker.state = "half-open"  # simulate cooldown elapsed
breaker.record_success()
print(f"After success: {breaker}")

Initial: CircuitBreaker(state=closed, failures=0/3)
After failure 1: CircuitBreaker(state=closed, failures=1/3)
After failure 2: CircuitBreaker(state=closed, failures=2/3)
After failure 3: CircuitBreaker(state=open, failures=3/3)
Can call? False
After success: CircuitBreaker(state=closed, failures=0/3)


In [96]:
# API call with circuit breaker integration

api_breaker = CircuitBreaker(failure_threshold=5, cooldown_seconds=30)

def call_with_circuit_breaker(
    messages: list[dict], max_tokens: int = 200, default_response: str = "[Service unavailable]"
) -> str:
    """Call API with circuit breaker protection. Returns default when circuit is open."""
    if not api_breaker.can_call():
        print(f"  Circuit OPEN — returning default (cooldown {api_breaker.cooldown_seconds}s)")
        return default_response
    content, error = safe_call(messages, max_tokens)
    if content:
        api_breaker.record_success()
        return content
    else:
        api_breaker.record_failure()
        print(f"  Failure recorded: {api_breaker}")
        return default_response

result = call_with_circuit_breaker([{"role": "user", "content": "Say OK"}])
print(f"Result: {result}")
print(f"Circuit state: {api_breaker}")

Result: OK!
Circuit state: CircuitBreaker(state=closed, failures=0/5)


### Graceful degradation

**Production pattern:** When the AI API is down, return a cached or default response instead of failing. In a batch pipeline, this means unsuccessful items get a fallback label that humans review later.

In [97]:
# In-memory cache for API responses (production: use Redis or database)
response_cache: dict[str, str] = {}

def call_with_cache(
    messages: list[dict], cache_key: str, max_tokens: int = 200
) -> tuple[str, str]:
    """Call API with cache fallback. Returns (content, source) where source is 'api' or 'cache'."""
    content, error = safe_call(messages, max_tokens)
    if content:
        response_cache[cache_key] = content
        return content, "api"
    # Fallback to cache
    if cache_key in response_cache:
        print(f"  API failed ({error}); using cached response")
        return response_cache[cache_key], "cache"
    return f"[Unavailable: {error}]", "fallback"

# Example: first call populates cache; subsequent calls can use it if API fails
msg = [{"role": "user", "content": "What is REST?"}]
result, source = call_with_cache(msg, cache_key="rest_definition")
print(f"Result ({source}): {result[:100]}")
print(f"Cache size: {len(response_cache)} entries")

Result (api): REST, which stands for Representational State Transfer, is an architectural style for designing netw
Cache size: 1 entries


## 7. Logging and debugging AI pipelines

**Real-life scenario:** Compliance requires a 90-day audit trail of all AI API calls with token usage, model versions, and input/output hashes. Every call must be traceable from the triggering event (e.g., ticket ID) through the API call to the final result. No `print()` in production — use structured logging.

### Log levels in practice

| Level | When to use | Example |
|-------|-----------|--------|
| DEBUG | Payload dumps, raw response (developer use) | `logger.debug(f"Payload: {json.dumps(payload)}")` |
| INFO | Request sent, response received, result saved | `logger.info(f"Summarized ticket #{id}")` |
| WARNING | Retry triggered, cache fallback used | `logger.warning(f"Retry 2/3 for ticket #{id}")` |
| ERROR | Call failed permanently, invalid response | `logger.error(f"Failed ticket #{id}: {error}")` |

In [98]:
import logging
import uuid

# Configure logging (production: JSON format to stdout → log aggregation system)
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s | %(levelname)-7s | %(name)s | %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)
logger = logging.getLogger("ai_pipeline")

print("Logger configured: ai_pipeline")
print("Levels: DEBUG, INFO, WARNING, ERROR")

Logger configured: ai_pipeline
Levels: DEBUG, INFO, WARNING, ERROR


In [99]:
# Correlation ID: trace a request from input through API call to result

def generate_request_id() -> str:
    """Generate a unique request ID for tracing."""
    return str(uuid.uuid4())[:8]


def summarize_with_logging(text: str, source_id: str = "") -> tuple[str | None, dict]:
    """Summarize text with full logging at every step."""
    request_id = generate_request_id()
    log_ctx = f"req={request_id} source={source_id}"

    logger.info(f"{log_ctx} | Starting summarization | input_length={len(text)}")

    if not API_KEY:
        logger.warning(f"{log_ctx} | API key not set — skipping")
        return "[No API key]", {"request_id": request_id}

    payload = build_chat_payload(
        "Output exactly 3 bullet points as executive summary.", text
    )
    logger.debug(f"{log_ctx} | Payload: {json.dumps(payload)[:200]}")

    start = time.monotonic()
    try:
        resp = call_chat_api(payload)
        elapsed_ms = (time.monotonic() - start) * 1000
        data = resp.json()
        content, meta = parse_chat_response(data)
        meta["request_id"] = request_id
        meta["elapsed_ms"] = round(elapsed_ms, 1)

        logger.info(
            f"{log_ctx} | Success | model={meta.get('model')} "
            f"tokens={meta.get('total_tokens')} elapsed={elapsed_ms:.0f}ms"
        )
        return content, meta

    except requests.exceptions.HTTPError as e:
        elapsed_ms = (time.monotonic() - start) * 1000
        status = e.response.status_code if e.response is not None else "unknown"
        logger.error(f"{log_ctx} | HTTP error {status} | elapsed={elapsed_ms:.0f}ms")
        return None, {"request_id": request_id, "error": f"HTTP {status}"}

    except requests.exceptions.Timeout:
        logger.error(f"{log_ctx} | Timeout after {TIMEOUT}s")
        return None, {"request_id": request_id, "error": "Timeout"}


summary, meta = summarize_with_logging(sample_report, source_id="ticket-QR-001")
print(f"\nResult: {summary[:150] if summary else 'None'}")
print(f"Trace: {meta}")

2026-03-10 19:07:48 | INFO    | ai_pipeline | req=c060b7d5 source=ticket-QR-001 | Starting summarization | input_length=542
2026-03-10 19:07:51 | INFO    | ai_pipeline | req=c060b7d5 source=ticket-QR-001 | Success | model=gpt-4o-mini-2024-07-18 tokens=272 elapsed=2688ms



Result: - Revenue increased by 12% year-over-year, primarily driven by heightened enterprise adoption and the signing of three new contracts totaling $2.4M AR
Trace: {'model': 'gpt-4o-mini-2024-07-18', 'finish_reason': 'stop', 'prompt_tokens': 156, 'completion_tokens': 116, 'total_tokens': 272, 'request_id': 'c060b7d5', 'elapsed_ms': 2688.0}


### Structured JSON logging

**Production pattern:** Log aggregation systems (ELK, Splunk, Datadog) parse JSON logs. Instead of free-text log lines, emit structured JSON with standard fields.

In [100]:
class JSONFormatter(logging.Formatter):
    """Emit log records as JSON lines for log aggregation systems."""
    def format(self, record: logging.LogRecord) -> str:
        log_entry = {
            "timestamp": self.formatTime(record),
            "level": record.levelname,
            "logger": record.name,
            "message": record.getMessage(),
        }
        return json.dumps(log_entry)

# Create a handler with JSON formatting (in notebook, print to stdout)
json_handler = logging.StreamHandler()
json_handler.setFormatter(JSONFormatter())

json_logger = logging.getLogger("ai_pipeline.json")
json_logger.addHandler(json_handler)
json_logger.setLevel(logging.INFO)

# Example log entries
json_logger.info("API call started | request_id=abc123 | model=gpt-4o-mini")
json_logger.info("API call completed | request_id=abc123 | tokens=47 | elapsed_ms=230")

{"timestamp": "2026-03-10 19:07:51,155", "level": "INFO", "logger": "ai_pipeline.json", "message": "API call started | request_id=abc123 | model=gpt-4o-mini"}
{"timestamp": "2026-03-10 19:07:51,155", "level": "INFO", "logger": "ai_pipeline.json", "message": "API call started | request_id=abc123 | model=gpt-4o-mini"}
2026-03-10 19:07:51 | INFO    | ai_pipeline.json | API call started | request_id=abc123 | model=gpt-4o-mini
{"timestamp": "2026-03-10 19:07:51,158", "level": "INFO", "logger": "ai_pipeline.json", "message": "API call completed | request_id=abc123 | tokens=47 | elapsed_ms=230"}
{"timestamp": "2026-03-10 19:07:51,158", "level": "INFO", "logger": "ai_pipeline.json", "message": "API call completed | request_id=abc123 | tokens=47 | elapsed_ms=230"}
2026-03-10 19:07:51 | INFO    | ai_pipeline.json | API call completed | request_id=abc123 | tokens=47 | elapsed_ms=230


### Performance logging and cost estimation

**Production pattern:** Track response time and token usage per call. Estimate cost (tokens * rate) to monitor spend.

In [101]:
# Token cost estimation (approximate rates as of 2024 — update for your provider)
COST_PER_1K_TOKENS = {
    "gpt-4o-mini": {"prompt": 0.00015, "completion": 0.0006},
    "gpt-4o": {"prompt": 0.005, "completion": 0.015},
}

def estimate_cost(model: str, prompt_tokens: int, completion_tokens: int) -> float:
    """Estimate API call cost in USD."""
    rates = COST_PER_1K_TOKENS.get(model, {"prompt": 0.001, "completion": 0.002})
    cost = (prompt_tokens / 1000 * rates["prompt"]) + (completion_tokens / 1000 * rates["completion"])
    return round(cost, 6)


def call_with_cost_tracking(
    messages: list[dict], max_tokens: int = 200
) -> dict:
    """Call API and return result with timing and cost estimate."""
    result = {"content": None, "elapsed_ms": 0, "tokens": {}, "estimated_cost_usd": 0}
    if not API_KEY:
        result["content"] = "[No API key]"
        return result

    start = time.monotonic()
    resp = call_chat_api(
        build_chat_payload("Be concise.", messages[-1]["content"], max_tokens=max_tokens)
    )
    result["elapsed_ms"] = round((time.monotonic() - start) * 1000, 1)

    data = resp.json()
    content, meta = parse_chat_response(data)
    result["content"] = content
    result["tokens"] = {
        "prompt": meta.get("prompt_tokens", 0),
        "completion": meta.get("completion_tokens", 0),
        "total": meta.get("total_tokens", 0),
    }
    result["estimated_cost_usd"] = estimate_cost(
        meta.get("model", MODEL_CHAT),
        result["tokens"]["prompt"],
        result["tokens"]["completion"],
    )
    return result


tracked = call_with_cost_tracking([{"role": "user", "content": "Explain REST in one sentence."}])
print(f"Content: {tracked['content']}")
print(f"Elapsed: {tracked['elapsed_ms']}ms")
print(f"Tokens: {tracked['tokens']}")
print(f"Estimated cost: ${tracked['estimated_cost_usd']}")

Content: REST (Representational State Transfer) is an architectural style for designing networked applications that uses standard HTTP methods to enable communication between clients and servers through stateless interactions and resource representations.
Elapsed: 2234.0ms
Tokens: {'prompt': 20, 'completion': 36, 'total': 56}
Estimated cost: $9.2e-05


### API key and PII redaction in logs

**Production rule:** Never log API keys, passwords, or PII. Redact before logging.

In [102]:
import re

def redact_sensitive(text: str) -> str:
    """Redact API keys and common PII patterns from log text."""
    # Redact Bearer tokens
    text = re.sub(r'Bearer\s+\S+', 'Bearer [REDACTED]', text)
    # Redact API key patterns (sk-...)
    text = re.sub(r'sk-[A-Za-z0-9]{20,}', '[REDACTED_KEY]', text)
    # Redact email addresses
    text = re.sub(r'[\w.+-]+@[\w-]+\.[\w.-]+', '[REDACTED_EMAIL]', text)
    return text


# Example: redact sensitive data before logging
raw_log = f"Authorization: Bearer {API_KEY or 'sk-test1234567890abcdefghijk'} | user: john.doe@company.com"
safe_log = redact_sensitive(raw_log)
print(f"Raw:     {raw_log[:80]}...")
print(f"Redacted: {safe_log}")

Raw:     Authorization: Bearer sk-proj-tnOIuM9ohh-ooN3W-gzsYGWA_JM2E-7QZzj2oUK4dX9NSqeeph...
Redacted: Authorization: Bearer [REDACTED] | user: [REDACTED_EMAIL]


### Audit trail pattern

**Real-life scenario:** Finance and compliance teams require a record of every AI-generated output — what went in, what came out, which model, when, and how many tokens. This audit log feeds into governance dashboards.

In [103]:
from datetime import datetime, timezone
import hashlib

audit_log: list[dict] = []

def record_audit_entry(
    request_id: str,
    source_id: str,
    input_text: str,
    output_text: str,
    model: str,
    tokens: int,
    elapsed_ms: float,
    cost_usd: float,
) -> dict:
    """Record an audit entry for compliance. Stores input/output hashes (not raw PII)."""
    entry = {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "request_id": request_id,
        "source_id": source_id,
        "model": model,
        "input_hash": hashlib.sha256(input_text.encode()).hexdigest()[:16],
        "output_length": len(output_text),
        "total_tokens": tokens,
        "elapsed_ms": elapsed_ms,
        "cost_usd": cost_usd,
    }
    audit_log.append(entry)
    return entry


# Record sample entry
entry = record_audit_entry(
    request_id="abc12345",
    source_id="ticket-42",
    input_text=sample_report,
    output_text=summary or "N/A",
    model=MODEL_CHAT,
    tokens=47,
    elapsed_ms=230.5,
    cost_usd=0.000028,
)
print("Audit entry:")
print(json.dumps(entry, indent=2))

Audit entry:
{
  "timestamp": "2026-03-10T13:37:53.469886+00:00",
  "request_id": "abc12345",
  "source_id": "ticket-42",
  "model": "gpt-4o-mini",
  "input_hash": "0e14d43355bc3809",
  "output_length": 545,
  "total_tokens": 47,
  "elapsed_ms": 230.5,
  "cost_usd": 2.8e-05
}


In [104]:
# Export audit log to CSV for compliance reporting

def export_audit_log(log: list[dict], output_path: str = "audit_log.csv") -> str:
    """Export audit log to CSV."""
    if not log:
        print("No audit entries to export.")
        return output_path
    fieldnames = list(log[0].keys())
    with open(output_path, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(log)
    print(f"Exported {len(log)} audit entries to {output_path}")
    return output_path

export_audit_log(audit_log)

Exported 1 audit entries to audit_log.csv


'audit_log.csv'

## 8. End-to-end production pipeline

**Real-life scenario:** Every night, a cron job runs your script to classify the day's incoming support tickets. It reads from CSV, classifies each ticket via the AI API with retry logic, logs every step with correlation IDs, and writes the augmented results to a new CSV for the operations team.

This section wires together everything from §3-§7: modular functions, CSV I/O, API calls with retry, structured logging, audit trail, and error handling.

**Maps to architecture:** Block A end-to-end — your code reads data, builds payloads, sends to Block B/C, parses responses, and persists results.

In [105]:
def run_classification_pipeline(
    input_csv: str = "tickets.csv",
    output_csv: str = "tickets_classified.csv",
    max_tickets: int | None = None,
) -> dict:
    """End-to-end pipeline: CSV → classify via API → write results CSV.
    Returns summary dict with counts and timing.
    """
    pipeline_id = generate_request_id()
    logger.info(f"pipeline={pipeline_id} | Starting classification pipeline | input={input_csv}")
    start = time.monotonic()

    # Step 1: Read input
    all_tickets = read_tickets(input_csv)
    if max_tickets:
        all_tickets = all_tickets[:max_tickets]
    logger.info(f"pipeline={pipeline_id} | Loaded {len(all_tickets)} tickets")

    # Step 2: Classify each ticket
    results = []
    success_count = 0
    error_count = 0
    total_tokens = 0

    for ticket in all_tickets:
        req_id = generate_request_id()
        logger.info(f"pipeline={pipeline_id} req={req_id} | Classifying ticket #{ticket['ticket_id']}")

        classified = classify_ticket(ticket)

        if classified["category"].startswith("[error") or classified["category"].startswith("[no_api"):
            error_count += 1
            logger.warning(f"pipeline={pipeline_id} req={req_id} | Failed: {classified['category']}")
        else:
            success_count += 1
            tokens = classified.get("tokens", 0) or 0
            total_tokens += tokens
            logger.info(
                f"pipeline={pipeline_id} req={req_id} | "
                f"Classified #{ticket['ticket_id']} → {classified['category']} | tokens={tokens}"
            )
        results.append(classified)

    # Step 3: Write results
    write_results(results, output_csv)

    elapsed = time.monotonic() - start
    summary = {
        "pipeline_id": pipeline_id,
        "total_tickets": len(all_tickets),
        "success": success_count,
        "errors": error_count,
        "total_tokens": total_tokens,
        "elapsed_seconds": round(elapsed, 2),
        "output_file": output_csv,
    }
    logger.info(f"pipeline={pipeline_id} | Complete | {json.dumps(summary)}")
    return summary

In [106]:
# Run the full pipeline on first 5 tickets
pipeline_summary = run_classification_pipeline(max_tickets=5)
print("\nPipeline summary:")
print(json.dumps(pipeline_summary, indent=2))

2026-03-10 19:07:53 | INFO    | ai_pipeline | pipeline=22fd213c | Starting classification pipeline | input=tickets.csv
2026-03-10 19:07:53 | INFO    | ai_pipeline | pipeline=22fd213c | Loaded 5 tickets
2026-03-10 19:07:53 | INFO    | ai_pipeline | pipeline=22fd213c req=adea6031 | Classifying ticket #1
2026-03-10 19:07:54 | INFO    | ai_pipeline | pipeline=22fd213c req=adea6031 | Classified #1 → technical | tokens=69
2026-03-10 19:07:54 | INFO    | ai_pipeline | pipeline=22fd213c req=39ed97e8 | Classifying ticket #2
2026-03-10 19:07:55 | INFO    | ai_pipeline | pipeline=22fd213c req=39ed97e8 | Classified #2 → billing | tokens=70
2026-03-10 19:07:55 | INFO    | ai_pipeline | pipeline=22fd213c req=5a2c0bcd | Classifying ticket #3
2026-03-10 19:07:56 | INFO    | ai_pipeline | pipeline=22fd213c req=5a2c0bcd | Classified #3 → feature_request | tokens=70
2026-03-10 19:07:56 | INFO    | ai_pipeline | pipeline=22fd213c req=636b6d38 | Classifying ticket #4
2026-03-10 19:07:57 | INFO    | ai_pipe

Wrote 5 rows to tickets_classified.csv

Pipeline summary:
{
  "pipeline_id": "22fd213c",
  "total_tickets": 5,
  "success": 5,
  "errors": 0,
  "total_tokens": 347,
  "elapsed_seconds": 7.56,
  "output_file": "tickets_classified.csv"
}


In [107]:
# Verify output: read back the classified CSV
if Path("tickets_classified.csv").exists():
    with open("tickets_classified.csv", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        print("Classified tickets output:")
        for row in reader:
            print(f"  #{row['ticket_id']} [{row.get('priority', '?')}] {row['subject'][:30]} → {row['category']}")

Classified tickets output:
  #1 [high] Login failed → technical
  #2 [high] Invoice discrepancy → billing
  #3 [medium] Export to PDF → feature_request
  #4 [high] API rate limit → technical
  #5 [low] Renewal date → billing


In [108]:
# Run full pipeline on ALL 18 tickets (production-like batch)
full_summary = run_classification_pipeline(
    input_csv="tickets.csv",
    output_csv="tickets_all_classified.csv",
)
print("\nFull batch pipeline summary:")
print(json.dumps(full_summary, indent=2))

2026-03-10 19:08:01 | INFO    | ai_pipeline | pipeline=04077fb7 | Starting classification pipeline | input=tickets.csv
2026-03-10 19:08:01 | INFO    | ai_pipeline | pipeline=04077fb7 | Loaded 18 tickets
2026-03-10 19:08:01 | INFO    | ai_pipeline | pipeline=04077fb7 req=85843b2c | Classifying ticket #1
2026-03-10 19:08:02 | INFO    | ai_pipeline | pipeline=04077fb7 req=85843b2c | Classified #1 → technical | tokens=69
2026-03-10 19:08:02 | INFO    | ai_pipeline | pipeline=04077fb7 req=4e324859 | Classifying ticket #2
2026-03-10 19:08:03 | INFO    | ai_pipeline | pipeline=04077fb7 req=4e324859 | Classified #2 → billing | tokens=70
2026-03-10 19:08:03 | INFO    | ai_pipeline | pipeline=04077fb7 req=1a7a8712 | Classifying ticket #3
2026-03-10 19:08:04 | INFO    | ai_pipeline | pipeline=04077fb7 req=1a7a8712 | Classified #3 → feature_request | tokens=70
2026-03-10 19:08:04 | INFO    | ai_pipeline | pipeline=04077fb7 req=3b157ff3 | Classifying ticket #4
2026-03-10 19:08:05 | INFO    | ai_pip

Wrote 18 rows to tickets_all_classified.csv

Full batch pipeline summary:
{
  "pipeline_id": "04077fb7",
  "total_tickets": 18,
  "success": 18,
  "errors": 0,
  "total_tokens": 1258,
  "elapsed_seconds": 19.58,
  "output_file": "tickets_all_classified.csv"
}


### Pipeline with support reply generation

**Real-life scenario:** After classifying tickets, generate draft support replies for the top-priority ones. The operations team reviews drafts before sending.

In [109]:
# Generate support reply drafts for high-priority tickets

GENERAL_KNOWLEDGE = """
Password reset: use Forgot password link. If error 500, clear cache or try incognito.
Billing: changes appear on next statement. For refunds, contact billing@company.com.
SSO/SAML: check IdP configuration. Verify ACS URL matches. Clear browser session.
Permissions: admin must grant workspace access. Check team roles in Settings > Members.
SSL/certificates: auto-renewal runs daily. For custom domains, re-upload certificate.
Webhooks: enable idempotency keys. Check endpoint health. Duplicates may occur during retries.
Storage: upgrade plan or archive old data. At 100% limit, uploads are blocked.
Search: reindex may be needed after bulk import. Check that documents are not in draft status.
"""

def generate_reply_for_ticket(ticket: dict) -> dict:
    """Generate a draft support reply for a ticket."""
    ticket_text = f"Subject: {ticket['subject']}\nBody: {ticket['body']}"
    reply = draft_support_reply(ticket_text, GENERAL_KNOWLEDGE)
    return {**ticket, "draft_reply": reply}

# Get high-priority tickets only
high_priority = [t for t in tickets if t.get("priority") == "high"]
print(f"High-priority tickets: {len(high_priority)}")

# Generate drafts for first 3
for t in high_priority[:3]:
    result = generate_reply_for_ticket(t)
    print(f"\n--- Ticket #{result['ticket_id']}: {result['subject']} ---")
    print(f"Draft reply: {result['draft_reply'][:200]}")

High-priority tickets: 8

--- Ticket #1: Login failed ---
Draft reply: Thank you for your message. For the error 500 when using the Forgot password link, please try clearing your browser cache or using an incognito window. If the issue persists, let us know.

--- Ticket #2: Invoice discrepancy ---
Draft reply: To address the invoice discrepancy, please note that changes will reflect on your next statement. For a refund, contact billing@company.com directly.

--- Ticket #4: API rate limit ---
Draft reply: To address the 429 rate limit errors, please review your API usage patterns and optimize unnecessary calls. Currently, limits cannot be increased. Consider implementing efficient polling or batching s


## 9. Frameworks quick reference

| Library | Purpose in this course | Key functions | Notes |
|---------|----------------------|---------------|-------|
| `requests` | HTTP calls to AI APIs | `post()`, `get()`, `Session()` | Main HTTP client; used in every API call |
| `json` | Build and parse payloads | `dumps()`, `loads()`, `load()` | Serialize Python dicts to JSON strings |
| `csv` | Read/write tabular data | `DictReader()`, `DictWriter()` | Ticket data, results export |
| `os` | Environment variables | `environ.get()` | API keys from .env |
| `logging` | Structured observability | `getLogger()`, `basicConfig()` | No print() in production |
| `pathlib` | File path handling | `Path()`, `.exists()`, `.read_bytes()` | Cross-platform paths |
| `uuid` | Unique identifiers | `uuid4()` | Request/correlation IDs |
| `time` | Performance measurement | `monotonic()`, `sleep()` | Timing and backoff delays |
| `dataclasses` | Typed config objects | `@dataclass`, `field()` | Type-safe configuration |
| `hashlib` | Content hashing | `sha256()` | Audit trail input hashes |

For the full cheat sheet with all frameworks used across the 3-day course, see `extra/01-Architecture-and-Frameworks-Reference.ipynb` (available on request; not covered in session).

In [110]:
# Runnable examples for each library

# --- requests ---
print("=== requests ===")
if API_KEY:
    r = requests.post(
        f"{API_BASE}/chat/completions",
        json={"model": MODEL_CHAT, "messages": [{"role": "user", "content": "Hi"}], "max_tokens": 5},
        headers={"Authorization": f"Bearer {API_KEY}", "Content-Type": "application/json"},
        timeout=30,
    )
    print(f"  Status: {r.status_code}, Content: {r.json()['choices'][0]['message']['content']}")
else:
    print("  requests.post(url, json=body, headers=headers, timeout=30)")

# --- json ---
print("\n=== json ===")
data = {"model": "gpt-4o-mini", "messages": [{"role": "user", "content": "test"}]}
json_str = json.dumps(data, indent=2)
parsed = json.loads(json_str)
print(f"  dumps: {json_str[:60]}...")
print(f"  loads: model={parsed['model']}")

# --- csv ---
print("\n=== csv ===")
import io
csv_text = "name,score\nAlice,95\nBob,87"
reader = csv.DictReader(io.StringIO(csv_text))
for row in reader:
    print(f"  {row['name']}: {row['score']}")

=== requests ===
  Status: 200, Content: Hello! How can I

=== json ===
  dumps: {
  "model": "gpt-4o-mini",
  "messages": [
    {
      "rol...
  loads: model=gpt-4o-mini

=== csv ===
  Alice: 95
  Bob: 87


In [111]:
# --- os ---
print("=== os ===")
print(f"  OPENAI_API_KEY set: {bool(os.environ.get('OPENAI_API_KEY'))}")
print(f"  HOME: {os.environ.get('HOME') or os.environ.get('USERPROFILE', 'N/A')}")

# --- logging ---
print("\n=== logging ===")
demo_logger = logging.getLogger("demo")
demo_logger.info("This is an INFO message with structured context | key=value")

# --- pathlib ---
print("\n=== pathlib ===")
p = Path("config.json")
print(f"  Path('config.json').exists(): {p.exists()}")
print(f"  Path.cwd(): {Path.cwd().name}")

# --- uuid ---
print("\n=== uuid ===")
print(f"  uuid4(): {uuid.uuid4()}")
print(f"  Short ID: {str(uuid.uuid4())[:8]}")

# --- time ---
print("\n=== time ===")
t0 = time.monotonic()
time.sleep(0.01)
print(f"  Elapsed: {(time.monotonic() - t0)*1000:.1f}ms")

# --- dataclasses ---
print("\n=== dataclasses ===")
print(f"  AIServiceConfig.from_env(): model={cfg.model_chat}, timeout={cfg.timeout}")

# --- hashlib ---
print("\n=== hashlib ===")
print(f"  sha256('hello'): {hashlib.sha256(b'hello').hexdigest()[:16]}")

2026-03-10 19:08:27 | INFO    | demo | This is an INFO message with structured context | key=value


=== os ===
  OPENAI_API_KEY set: True
  HOME: C:\Users\PC

=== logging ===

=== pathlib ===
  Path('config.json').exists(): True
  Path.cwd(): day-1

=== uuid ===
  uuid4(): 13808e42-9821-47b4-a08d-83a53be29747
  Short ID: b7e2dd9c

=== time ===
  Elapsed: 16.0ms

=== dataclasses ===
  AIServiceConfig.from_env(): model=gpt-4o-mini, timeout=60

=== hashlib ===
  sha256('hello'): 2cf24dba5fb0a30e


## Summary

- **Modular design**: Separate `build_payload` → `call_api` → `parse_response` functions. Each is testable and reusable. Config and secrets loaded once at startup.
- **REST pattern**: Every AI API call is `POST` with JSON body + auth header → parse JSON response. Same pattern for chat, STT, TTS, vision.
- **Data handling**: CSV → dict (DictReader) → API payload → results CSV. Validate payloads before sending. Parse responses defensively.
- **Error handling**: Classify errors by status code. Retry transient errors (429, 5xx) with exponential backoff and jitter. Circuit breaker for persistent failures. Cache fallback for graceful degradation.
- **Logging**: Structured logging with correlation IDs, timing, and token counts. Redact secrets and PII. Audit trail for compliance.
- **Production pipeline**: End-to-end CSV → classify → write results, with logging and error handling at every step.

**Next:** Module 02 (Multimodal AI Integration) applies these same patterns to text, image, and speech APIs.